File: backend/app/core/banks.py
"""
Curated Indian Bank Registry and Bank-Specific IFSC Code Validation.

Contains standard IFSC prefixes, bank metadata, and strict format validation rules.
"""

import re
from typing import TypedDict


class BankInfo(TypedDict):
    name: str
    code: str
    ifsc_prefix: str
    example_ifsc: str


SUPPORTED_BANKS: list[BankInfo] = [
    {
        "name": "State Bank of India",
        "code": "SBI",
        "ifsc_prefix": "SBIN",
        "example_ifsc": "SBIN0001234",
    },
    {
        "name": "HDFC Bank",
        "code": "HDFC",
        "ifsc_prefix": "HDFC",
        "example_ifsc": "HDFC0001234",
    },
    {
        "name": "ICICI Bank",
        "code": "ICICI",
        "ifsc_prefix": "ICIC",
        "example_ifsc": "ICIC0001234",
    },
    {
        "name": "Axis Bank",
        "code": "AXIS",
        "ifsc_prefix": "UTIB",
        "example_ifsc": "UTIB0001234",
    },
    {
        "name": "Kotak Mahindra Bank",
        "code": "KOTAK",
        "ifsc_prefix": "KKBK",
        "example_ifsc": "KKBK0001234",
    },
    {
        "name": "Punjab National Bank",
        "code": "PNB",
        "ifsc_prefix": "PUNB",
        "example_ifsc": "PUNB0001234",
    },
    {
        "name": "Bank of Baroda",
        "code": "BOB",
        "ifsc_prefix": "BARB",
        "example_ifsc": "BARB0001234",
    },
    {
        "name": "Canara Bank",
        "code": "CANARA",
        "ifsc_prefix": "CNRB",
        "example_ifsc": "CNRB0001234",
    },
    {
        "name": "Union Bank of India",
        "code": "UNION",
        "ifsc_prefix": "UBIN",
        "example_ifsc": "UBIN0001234",
    },
    {
        "name": "Indian Bank",
        "code": "INDIAN",
        "ifsc_prefix": "IDIB",
        "example_ifsc": "IDIB0001234",
    },
    {
        "name": "IDBI Bank",
        "code": "IDBI",
        "ifsc_prefix": "IBKL",
        "example_ifsc": "IBKL0001234",
    },
    {
        "name": "IndusInd Bank",
        "code": "INDUSIND",
        "ifsc_prefix": "INDB",
        "example_ifsc": "INDB0001234",
    },
    {
        "name": "Federal Bank",
        "code": "FEDERAL",
        "ifsc_prefix": "FDRL",
        "example_ifsc": "FDRL0001234",
    },
    {
        "name": "Yes Bank",
        "code": "YES",
        "ifsc_prefix": "YESB",
        "example_ifsc": "YESB0001234",
    },
    {
        "name": "Bank of India",
        "code": "BOI",
        "ifsc_prefix": "BKID",
        "example_ifsc": "BKID0001234",
    },
]

# Quick lookup maps
BANK_BY_NAME = {b["name"].lower(): b for b in SUPPORTED_BANKS}
BANK_BY_CODE = {b["code"].upper(): b for b in SUPPORTED_BANKS}

IFSC_REGEX = re.compile(r"^[A-Z]{4}0[A-Z0-9]{6}$")


def get_supported_banks() -> list[BankInfo]:
    """
    Returns the list of configured supported Indian banks.
    """
    return SUPPORTED_BANKS


def validate_bank_and_ifsc(bank_name: str, ifsc_code: str) -> tuple[bool, str | None]:
    """
    Validates general IFSC format AND verifies that the IFSC prefix matches the chosen bank.
    Returns (is_valid: bool, error_message: str | None).
    """
    clean_ifsc = (ifsc_code or "").strip().upper()
    clean_bank = (bank_name or "").strip()

    if not clean_ifsc:
        return False, "IFSC code is required."

    if not IFSC_REGEX.match(clean_ifsc):
        return False, "Invalid IFSC format. IFSC must be 11 alphanumeric characters (e.g. HDFC0001234, 5th character must be '0')."

    # Find bank in registry
    matched_bank = BANK_BY_NAME.get(clean_bank.lower())
    if not matched_bank:
        # Check by code or partial name
        for b in SUPPORTED_BANKS:
            if b["code"].lower() in clean_bank.lower() or clean_bank.lower() in b["name"].lower():
                matched_bank = b
                break

    if matched_bank:
        expected_prefix = matched_bank["ifsc_prefix"]
        if not clean_ifsc.startswith(expected_prefix):
            return False, (
                f"This IFSC code does not match {matched_bank['name']}. "
                f"Expected prefix '{expected_prefix}' (e.g. {matched_bank['example_ifsc']})."
            )

    return True, None

File: frontend/src/lib/banks.ts
/**
 * Curated Indian Bank Registry & IFSC prefix configuration.
 */

export interface BankConfig {
  name: string;
  code: string;
  ifscPrefix: string;
  exampleIfsc: string;
}

export const SUPPORTED_BANKS: BankConfig[] = [
  {
    name: 'State Bank of India',
    code: 'SBI',
    ifscPrefix: 'SBIN',
    exampleIfsc: 'SBIN0001234',
  },
  {
    name: 'HDFC Bank',
    code: 'HDFC',
    ifscPrefix: 'HDFC',
    exampleIfsc: 'HDFC0001234',
  },
  {
    name: 'ICICI Bank',
    code: 'ICICI',
    ifscPrefix: 'ICIC',
    exampleIfsc: 'ICIC0001234',
  },
  {
    name: 'Axis Bank',
    code: 'AXIS',
    ifscPrefix: 'UTIB',
    exampleIfsc: 'UTIB0001234',
  },
  {
    name: 'Kotak Mahindra Bank',
    code: 'KOTAK',
    ifscPrefix: 'KKBK',
    exampleIfsc: 'KKBK0001234',
  },
  {
    name: 'Punjab National Bank',
    code: 'PNB',
    ifscPrefix: 'PUNB',
    exampleIfsc: 'PUNB0001234',
  },
  {
    name: 'Bank of Baroda',
    code: 'BOB',
    ifscPrefix: 'BARB',
    exampleIfsc: 'BARB0001234',
  },
  {
    name: 'Canara Bank',
    code: 'CANARA',
    ifscPrefix: 'CNRB',
    exampleIfsc: 'CNRB0001234',
  },
  {
    name: 'Union Bank of India',
    code: 'UNION',
    ifscPrefix: 'UBIN',
    exampleIfsc: 'UBIN0001234',
  },
  {
    name: 'Indian Bank',
    code: 'INDIAN',
    ifscPrefix: 'IDIB',
    exampleIfsc: 'IDIB0001234',
  },
  {
    name: 'IDBI Bank',
    code: 'IDBI',
    ifscPrefix: 'IBKL',
    exampleIfsc: 'IBKL0001234',
  },
  {
    name: 'IndusInd Bank',
    code: 'INDUSIND',
    ifscPrefix: 'INDB',
    exampleIfsc: 'INDB0001234',
  },
  {
    name: 'Federal Bank',
    code: 'FEDERAL',
    ifscPrefix: 'FDRL',
    exampleIfsc: 'FDRL0001234',
  },
  {
    name: 'Yes Bank',
    code: 'YES',
    ifscPrefix: 'YESB',
    exampleIfsc: 'YESB0001234',
  },
  {
    name: 'Bank of India',
    code: 'BOI',
    ifscPrefix: 'BKID',
    exampleIfsc: 'BKID0001234',
  },
];

export const IFSC_REGEX = /^[A-Z]{4}0[A-Z0-9]{6}$/;

export function validateBankIfsc(bankName: string, ifsc: string): { isValid: boolean; message: string } {
  const cleanIfsc = (ifsc || '').trim().toUpperCase();
  const cleanBank = (bankName || '').trim();

  if (!cleanIfsc) {
    return { isValid: false, message: 'IFSC code is required.' };
  }

  if (!IFSC_REGEX.test(cleanIfsc)) {
    return {
      isValid: false,
      message: 'Invalid IFSC format. Must be 11 characters (4 letters, 0, 6 characters, e.g. HDFC0001234).',
    };
  }

  const foundBank = SUPPORTED_BANKS.find(
    (b) =>
      b.name.toLowerCase() === cleanBank.toLowerCase() ||
      b.code.toLowerCase() === cleanBank.toLowerCase()
  );

  if (foundBank) {
    if (!cleanIfsc.startsWith(foundBank.ifscPrefix)) {
      return {
        isValid: false,
        message: `⚠ This IFSC code does not match ${foundBank.name}. ${foundBank.name} IFSC must start with "${foundBank.ifscPrefix}" (e.g. ${foundBank.exampleIfsc}).`,
      };
    }
  }

  return { isValid: true, message: '✓ IFSC format is valid for selected bank' };
}
